### PR TITLE
Add input validation to query.GetVirtualChannel

### PR DESCRIPTION
**Pull request by bitwiseguy**
In order to reduce friction of go-nitro integration into boost, this PR adds some input validation to the rpcClient.GetVirtualChannel method. This means the validation occurs inside go-nitro code instead of the burden being placed on boost or another consumer

Resolves https://github.com/statechannels/boost/issues/8

### CODE REVIEW

File: rpc/client.go [hunk 1]
```diff
@@ -80,13 +82,16 @@ func NewHttpRpcClient(rpcServerUrl string) (*RpcClient, error) {
...
-func (rc *RpcClient) GetVirtualChannel(id types.Destination) query.PaymentChannelInfo {
-	req := serde.GetPaymentChannelRequest{Id: id}
+func (rc *RpcClient) GetVirtualChannel(chId types.Destination) (query.PaymentChannelInfo, int, error) {
+	if (chId == types.Destination{}) {
+		return query.PaymentChannelInfo{}, http.StatusBadRequest, errors.New("a valid channel id must be provided")
+	}
+	req := serde.GetPaymentChannelRequest{Id: chId}
 
-	return waitForRequest[serde.GetPaymentChannelRequest, query.PaymentChannelInfo](rc, serde.GetPaymentChannelRequestMethod, req)
+	return waitForRequest[serde.GetPaymentChannelRequest, query.PaymentChannelInfo](rc, serde.GetPaymentChannelRequestMethod, req), http.StatusOK, nil
```
Comment (geoknee) [inline rpc/client.go:91]: I'm not sure about returning an http status code. Although in practice we tend to use http as a transport for the RPC connection, it is in fact transport agnostic.

If the input validation were being performed by the RPC server (instead of the client), it could make more sense to return an http error (again depending on the transport). 

In fact, thinking about this some more, I wonder if it does in fact make more sense to the validation on the server side.

The main thrust behind having validation is to avoid the nitro node panicking https://github.com/statechannels/boost/pull/3#discussion_r1223402151 . 

The deeper in we put the validation, the more protection we get. Putting the validation in the go rpc client means we don't get protection when
* someone uses the TS rpc client (unless we duplicate the validation there too)
* someone uses go-nitro as a library, not as a service

So I think we should rethink our approach a little. Perhaps start with a test which triggers the go-nitro node to panic by sending in a query for a zero channel using the rpc client. Then, we would like to transform that panic to an error, bubble it up to the rpc sever, and send it back to the rpc client as an json rpc error response. Ultimately that will be received by the rpc client and it will still cause an error to be returned from `RpcClient.GetVirtualChannel`.

Keen to hear your thoughts! And apologies that the direction wasn't clear from the outset.

Comment (bitwiseguy) [reply rpc/client.go:91]: That makes sense to me and I agree that pushing validation deeper would be better. I noticed that in our jsonrpc server.go we have methods for `GetPaymentChannel` and `GetLedgerChannel`. Would `GetVirtualChannel` be the same as `GetPaymentChannel` or should I create an additional jsonrpc method?

Comment (bitwiseguy) [reply rpc/client.go:91]: There are a few other instances in the rpc client code where "Virtual" is used instead of "Payment". To improve the naming consistency, would you agree with renaming the following client methods?

`CreateVirtual` --> `CreatePaymentChannel`
`CloseVirtual` --> `ClosePaymentChannel`
`CreateLedger` --> `CreateLedgerChannel`
`CloseLedger` --> `CloseLedgerChannel`

Comment (bitwiseguy) [reply rpc/client.go:91]: I wonder if we might also want to align some of the json-rpc methods with the names of the core/node methods. The following method names are a little confusing:

json-rpc | node
-- | --
virtual_fund | CreateVirtualPaymentChannel
virtual_defund | CloseVirtualChannel
direct_fund | CreateLedgerChannel
direct_defund | CloseLedgerChannel

File: rpc/client.go [hunk 2]
```diff
@@ -80,14 +79,14 @@ func NewHttpRpcClient(rpcServerUrl string) (*RpcClient, error) {
...
-	req := serde.GetPaymentChannelRequest{Id: id}
+func (rc *RpcClient) GetPaymentChannel(chId types.Destination) query.PaymentChannelInfo {
+	req := serde.GetPaymentChannelRequest{Id: chId}
 
 	return waitForRequest[serde.GetPaymentChannelRequest, query.PaymentChannelInfo](rc, serde.GetPaymentChannelRequestMethod, req)
 }
 
-// CreateLedger creates a new ledger channel
-func (rc *RpcClient) CreateVirtual(intermediaries []types.Address, counterparty types.Address, ChallengeDuration uint32, outcome outcome.Exit) virtualfund.ObjectiveResponse {
+// CreatePaymentChannel creates a new virtual channel
```
Comment (geoknee) [inline rpc/client.go:88]:
```suggestion
// CreatePaymentChannel creates a new virtual payment channel
```

File: rpc/client.go [hunk 3]
```diff
@@ -96,15 +95,15 @@ func (rc *RpcClient) CreateVirtual(intermediaries []types.Address, counterparty
 		rand.Uint64(),
 		common.Address{})
 
-	return waitForRequest[virtualfund.ObjectiveRequest, virtualfund.ObjectiveResponse](rc, serde.VirtualFundRequestMethod, objReq)
+	return waitForRequest[virtualfund.ObjectiveRequest, virtualfund.ObjectiveResponse](rc, serde.CreatePaymentChannelRequestMethod, objReq)
 }
 
-// CloseVirtual closes a virtual channel
-func (rc *RpcClient) CloseVirtual(id types.Destination) protocols.ObjectiveId {
+// ClosePaymentChannel closes a virtual channel
```
Comment (geoknee) [inline rpc/client.go:101]:
```suggestion
// ClosePaymentChannel attempts to close the payment channel with supplied id
```

When we get on to working on our next grant, we should be able to drop the "attempts to" disclaimer. For now, closing a channel requires cooperative counterparties.

File: rpc/client.go [hunk 4]
```diff
@@ -124,29 +123,29 @@ func (rc *RpcClient) GetPaymentChannelsByLedger(ledgerId types.Destination) []qu
...
+	return waitForRequest[directdefund.ObjectiveRequest, protocols.ObjectiveId](rc, serde.CloseLedgerChannelRequestMethod, objReq)
 }
 
 // Pay uses the specified channel to pay the specified amount
-func (rc *RpcClient) Pay(id types.Destination, amount uint64) {
+func (rc *RpcClient) Pay(id types.Destination, amount uint64) serde.PaymentRequest {
 	pReq := serde.PaymentRequest{Amount: amount, Channel: id}
 
-	waitForRequest[serde.PaymentRequest, serde.PaymentRequest](rc, serde.PayRequestMethod, pReq)
+	return waitForRequest[serde.PaymentRequest, serde.PaymentRequest](rc, serde.PayRequestMethod, pReq)
```
Comment (geoknee) [inline rpc/client.go:148]: Is this change intended / relevant?

File: rpc/serde/jsonrpc.go [hunk 5]
```diff
@@ -13,17 +13,17 @@ import (
...
-	GetPaymentChannelRequestMethod   RequestMethod = "get_payment_channel"
-	GetLedgerChannelRequestMethod    RequestMethod = "get_ledger_channel"
-	GetPaymentChannelsByLedgerMethod RequestMethod = "get_payment_channels_by_ledger"
-	GetAllLedgerChannelsMethod       RequestMethod = "get_all_ledger_channels"
+	GetAddressMethod                  RequestMethod = "get_address"
+	VersionMethod                     RequestMethod = "version"
+	CreateLedgerChannelRequestMethod  RequestMethod = "create_ledger_channel"
+	CloseLedgerChannelRequestMethod   RequestMethod = "close_ledger_channel"
+	CreatePaymentChannelRequestMethod RequestMethod = "create_payment_channel"
+	ClosePaymentChannelRequestMethod  RequestMethod = "close_payment_channel"
```
Comment (geoknee) [inline rpc/serde/jsonrpc.go:21]: 👍

File: rpc/client.go [hunk 6]
```diff
@@ -80,14 +79,14 @@ func NewHttpRpcClient(rpcServerUrl string) (*RpcClient, error) {
 	return c, nil
 }
 
-func (rc *RpcClient) GetVirtualChannel(id types.Destination) query.PaymentChannelInfo {
-	req := serde.GetPaymentChannelRequest{Id: id}
+func (rc *RpcClient) GetPaymentChannel(chId types.Destination) query.PaymentChannelInfo {
+	req := serde.GetPaymentChannelRequest{Id: chId}
 
 	return waitForRequest[serde.GetPaymentChannelRequest, query.PaymentChannelInfo](rc, serde.GetPaymentChannelRequestMethod, req)
```
Comment (geoknee) [inline rpc/client.go:85]: I'm not 100% sure but it looks like `waitForRequest` will panic in many error conditions, such that we will still crash the go-nitro server if we query for a channel with.a zero id. So still some more work to do to fix that! https://github.com/statechannels/go-nitro/issues/1275.

**Review by geoknee — APPROVED** [review]
Looks good!

**Comment by geoknee** [comment]
@bitwiseguy I put the breaking change label on this, I think we will need to update TS rpc client in tandem with this change.

File: rpc/client.go (hunk 4 above)
Comment (bitwiseguy) [reply rpc/client.go:148]: Reverted

File: rpc/client.go (hunk 6 above)
Comment (bitwiseguy) [reply rpc/client.go:85]: I added [this line](https://github.com/statechannels/go-nitro/pull/1368/files#diff-e411c027b510060c0ba3026138d11ece9c213a5e5927164270fb1f9ed45a8249R180) to test that the server does not panic when passed a non-existent chId. I believe its working as expected.

**Comment by bitwiseguy** [comment]
> @bitwiseguy I put the breaking change label on this, I think we will need to update TS rpc client in tandem with this change.

I will work on a PR for the TS rpc client and queue that one up before merging this one

File: rpc/client.go (hunk 6 above)
Comment (geoknee) [reply rpc/client.go:85]: Awesome, I guess that settles it then 👍

**Comment by bitwiseguy** [comment]
PR to make similar updates to TS client: https://github.com/statechannels/nitro-gui/pull/79
